cmake/modules/boards.cmake

```diff
@@ -94,7 +94,11 @@ endif()
 include(${ZEPHYR_BASE}/boards/deprecated.cmake)
 if(${BOARD}_DEPRECATED)
   set(BOARD_DEPRECATED ${BOARD} CACHE STRING "Deprecated board name, provided by user")
-  set(BOARD ${${BOARD}_DEPRECATED})
+  parse_board_components(${BOARD}_DEPRECATED BOARD BOARD_DEPRECATED_REVISION BOARD_DEPRECATED_IDENTIFIER)
+  if(NOT DEFINED BOARD_REVISION)
+    set(BOARD_REVISION ${BOARD_DEPRECATED_REVISION})
+  endif()
+  set(BOARD_IDENTIFIER ${BOARD_DEPRECATED_IDENTIFIER}${BOARD_IDENTIFIER})
```

Comment:
is this solving the right problem ?

A legacy board will not use the identifier form, so for those the code will work, such as:
```
set(nrf9160dk_nrf9160_ns_DEPRECATED nrf9160dk/nrf9160/ns)
set(nrf9160dk_nrf9160_DEPRECATED nrf9160dk/nrf9160)
```
Also, boards in v1 format should not be allowed to mix in `/`.

As example, deprecating the two nrf9160dk variants, be done as described above, however with this proposal, it suddenly becomes possible to mix in `/` in legacy board format, for example (mix of old and new: `nrf9160dk_nrf9160/ns`):
```
$ cmake -GNinja -DBOARD=nrf9160dk_nrf9160/ns -Bbuild samples/hello_world/
Loading Zephyr default modules (Zephyr repository).
-- Application: /projects/github/ncs/zephyr/samples/hello_world
CMake Warning at /projects/github/ncs/zephyr/cmake/modules/boards.cmake:102 (message):
  Deprecated BOARD=nrf9160dk_nrf9160 name specified, board automatically
  changed to: nrf9160dk.
...

-- Board: nrf9160dk, Revision: 0.14.0, identifier: nrf9160/ns
```
This should not have been allowed. If users want to specify new format (`/`), then they should also use the new name.

In future, what if someone is to deprecate a v2 board, either because of a SoC name change or a variant change, for example:
```
set(plank/foo_DEPRECATED plank/bar)              # SoC foo renamed to SoC bar.
set(plank/foo/nonsecure_DEPRECATED plank/foo/ns) # Variant rename.
```
Example of a variant rename in the past: https://github.com/zephyrproject-rtos/zephyr/commit/ccc595836a11b1d9146b0a25687192f9090eecdf.

afaict, such deprecation will not work because the board provided by the user has already been split.
https://github.com/zephyrproject-rtos/zephyr/blob/6c4a8bf749015ba9b703bbb42a18a190c79d3eda/cmake/modules/boards.cmake#L79

Reply:
> boards in v1 format should not be allowed to mix in `/`.

I understand you're apprehensive towards concatenating identifiers in general, but let's assume we go with that.

I agree that users shouldn't mix `/` with v1 board names, but I also think they shouldn't be using deprecated names in the first place. To me, putting restrictions on the use of `/` seems needlessly pedantic, especially if it could hinder any valid use cases with v2 board deprecation.

That said, we could perhaps apply different rules when deprecating v1 and v2 boards, if we added an extra flag for that. Personally though, I wouldn't be fond of having it in `deprecated.cmake`, like:
```
set(nrf9160dk_nrf9160_DEPRECATED nrf9160dk/nrf9160)
set(nrf9160dk_nrf9160_DEPRECATED_HWM v1)
```
This would look more presentable in yet another YAML file - something that we could consider anyway if we want deprecated boards to be recognized by twister, but that's another story.

> In future, what if someone is to deprecate a v2 board, either because of a SoC name change or a variant change, for example:
> 
> ```
> set(plank/foo_DEPRECATED plank/bar)              # SoC foo renamed to SoC bar.
> set(plank/foo/nonsecure_DEPRECATED plank/foo/ns) # Variant rename.
> ```

That's something I have consciously left out. It's not clear to me how deprecating any part of the platform identifier should work, whether it's a board revision, SoC name, CPU name, or other. The current implementation supports deprecating board names only.

This could be another thing that's better to express in YAML.

Reply:
> That said, we could perhaps apply different rules when deprecating v1 and v2 boards, if we added an extra flag for that. Personally though, I wouldn't be fond of having it in deprecated.cmake, like:

if having to explicitly deprecate board and board identifiers in deprecated to new names then this would really not be any issue.
For example:
```
set(plank/foo_DEPRECATED plank/bar)              # SoC foo renamed to SoC bar.
set(plank/foo/nonsecure_DEPRECATED plank/foo/ns) # Variant rename.
```
then building for `plank/foo/nonsecure` will map to `plank/foo/ns`, and checked like that.
`plank/foo/nonsecure/bar` will not match an entry and thus fail as invalid.

Question then becomes, how to handle a pure board rename, where you don't want to create an entry for each identifier, like a `plank` --> `plonk` rename.

That could be handled as:
```
set(plank_DEPRECATED plonk) # Pure rename on board level, any identifier applied by user will be transfered.
```
So that `plank/foo` --> `plonk/foo`.

whereas:
```
set(plank_foo_DEPRECATED plank/foo) # Board to board/soc rename, user usage of `/` is not allowed in -DBOARD.
```
So `plank_foo` --> `plank/foo`.
and `plank_foo/bar` is not allowed --> invalid board.

```
set(plank/foo_DEPRECATED plank/bar) # An identifier rename, each identifier must have an entry.
set(plank/foo/nonsecure_DEPRECATED plank/bar/ns)
```
So `plank/foo` --> `plank/bar`.
and `plank/foo/nonsecure` --> `plank/bar/ns`.
and `plank/foo/baz` is not mapped as it doesn't have an explicit entry (expected to result in invalid board / identifier).

This will allow for a pure board rename.
It will support the v1 to v2 deprecation of legacy boards into new names.
It will support variant renaming, and the number of variants on a given board should be maintainable. In principle it's no different from what we have in zephyr/main today where renaming a SoC or variant like `nonsecure` -> `ns` requires an entry per board. (each variant in legacy is a board)

Reply:
this can be handled depending on whether or not `/` is present in either deprecated or new part of name.

Reply:
Thanks for your insights.

I don't think we're ready to support deprecating v2 boards yet. I just pushed an update which limits this feature to v1 boards, ensures that v1 -> v2 transitions work, and prevents `/` from being mixed in inappropriately.

---
My main issue with this idea:
> ```
> set(plank/foo_DEPRECATED plank/bar)              # SoC foo renamed to SoC bar.
> ```
is that it's not a scalable way to deprecate a SoC name, which might be used by multiple boards in and out of tree.
```cmake
  set(bl5340_dvk/nrf5340_DEPRECATED                     bl5340_dvk/nrf_5340)
# set(bl5340_dvk/nrf5340/cpuapp_DEPRECATED              bl5340_dvk/nrf_5340/cpuapp)
# set(bl5340_dvk/nrf5340/cpuapp/ns_DEPRECATED           bl5340_dvk/nrf_5340/cpuapp/ns)
# set(bl5340_dvk/nrf5340/cpunet_DEPRECATED              bl5340_dvk/nrf_5340/cpunet)
  set(nrf5340_audio_dk/nrf5340_DEPRECATED               nrf5340_audio_dk/nrf_5340)
# set(nrf5340_audio_dk/nrf5340/cpuapp_DEPRECATED        nrf5340_audio_dk/nrf_5340/cpuapp)
# set(nrf5340_audio_dk/nrf5340/cpuapp/ns_DEPRECATED     nrf5340_audio_dk/nrf_5340/cpuapp/ns)
# set(nrf5340_audio_dk/nrf5340/cpunet_DEPRECATED        nrf5340_audio_dk/nrf_5340/cpunet)
  set(nrf5340dk/nrf5340_DEPRECATED                      nrf5340dk/nrf_5340)
# set(nrf5340dk/nrf5340/cpuapp_DEPRECATED               nrf5340dk/nrf_5340/cpuapp)
# set(nrf5340dk/nrf5340/cpuapp/ns_DEPRECATED            nrf5340dk/nrf_5340/cpuapp/ns)
# set(nrf5340dk/nrf5340/cpunet_DEPRECATED               nrf5340dk/nrf_5340/cpunet)
  set(pan1783_evb/nrf5340_DEPRECATED                    pan1783_evb/nrf_5340)
# set(pan1783_evb/nrf5340/cpuapp_DEPRECATED             pan1783_evb/nrf_5340/cpuapp)
# set(pan1783_evb/nrf5340/cpunet_DEPRECATED             pan1783_evb/nrf_5340/cpunet)
  set(pan1783a_evb/nrf5340_DEPRECATED                   pan1783a_evb/nrf_5340)
# set(pan1783a_evb/nrf5340/cpuapp_DEPRECATED            pan1783a_evb/nrf_5340/cpuapp)
# set(pan1783a_evb/nrf5340/cpunet_DEPRECATED            pan1783a_evb/nrf_5340/cpunet)
  set(pan1783a_pa_evb/nrf5340_DEPRECATED                pan1783a_pa_evb/nrf_5340)
# set(pan1783a_pa_evb/nrf5340/cpuapp_DEPRECATED         pan1783a_pa_evb/nrf_5340/cpuapp)
# set(pan1783a_pa_evb/nrf5340/cpunet_DEPRECATED         pan1783a_pa_evb/nrf_5340/cpunet)
  set(raytac_mdbt53_db_40/nrf5340_DEPRECATED            raytac_mdbt53_db_40/nrf_5340)
# set(raytac_mdbt53_db_40/nrf5340/cpuapp_DEPRECATED     raytac_mdbt53_db_40/nrf_5340/cpuapp)
# set(raytac_mdbt53_db_40/nrf5340/cpuapp/ns_DEPRECATED  raytac_mdbt53_db_40/nrf_5340/cpuapp/ns)
# set(raytac_mdbt53_db_40/nrf5340/cpunet_DEPRECATED     raytac_mdbt53_db_40/nrf_5340/cpunet)
  set(raytac_mdbt53v_db_40/nrf5340_DEPRECATED           raytac_mdbt53v_db_40/nrf_5340)
# set(raytac_mdbt53v_db_40/nrf5340/cpuapp_DEPRECATED    raytac_mdbt53v_db_40/nrf_5340/cpuapp)
# set(raytac_mdbt53v_db_40/nrf5340/cpuapp/ns_DEPRECATED raytac_mdbt53v_db_40/nrf_5340/cpuapp/ns)
# set(raytac_mdbt53v_db_40/nrf5340/cpunet_DEPRECATED    raytac_mdbt53v_db_40/nrf_5340/cpunet)
  set(thingy53/nrf5340_DEPRECATED                       thingy53/nrf_5340)
# set(thingy53/nrf5340/cpuapp_DEPRECATED                thingy53/nrf_5340/cpuapp)
# set(thingy53/nrf5340/cpuapp/ns_DEPRECATED             thingy53/nrf_5340/cpuapp/ns)
# set(thingy53/nrf5340/cpunet_DEPRECATED                thingy53/nrf_5340/cpunet)
```
Some lines are commented out, depending on which statements one might consider redundant.

You see, my first impression of `set(plank/foo_DEPRECATED plank/bar)` was that it could be a plain renaming rule: `plank/{foo -> bar}/**`, where any variant of `foo` would be assumed to still exist under `bar`.

To be fair, board definitions can change such that `plank/{foo -> bar}` doesn't imply `plank/{foo -> bar}/ns`, because `plank/bar/ns` doesn't exist anymore. However, we could have a different way of indicating this:
```
set(plank/foo_DEPRECATED    plank/bar)
set(plank/foo/ns_DEPRECATED NOTFOUND)
```
But this is just a matter of intuitiveness. In both cases, the approach still has obvious shortcomings.

> Question then becomes, how to handle a pure board rename, where you don't want to create an entry for each identifier, like a `plank` --> `plonk` rename.

If it were consistent with your overall approach, it would have to be like this:
```cmake
set(plank_DEPRECATED             plonk)
set(plank/soc_DEPRECATED         plonk/soc)
set(plank/soc/cpu0_DEPRECATED    plonk/soc/cpu0)
set(plank/soc/cpu0/ns_DEPRECATED plonk/soc/cpu0/ns)
set(plank/soc/cpu1_DEPRECATED    plonk/soc/cpu1)
```
I think there's no reason to treat the top level identifier differently from any other level.

---
I don't have a better proposal for v2 board deprecation at the moment, but I'd prefer it to properly account for the interdependence of boards and SoCs (which will grow in complexity).

Reply:
> is that it's not a scalable way to deprecate a SoC name, which might be used by multiple boards in and out of tree.

For a SoC rename, correct.
But the same can be said for v1 mode.

If a SoC like nrf5340 was renamed in zephyr/main, then all the boards in main featuring nrf5340 SoC in their board name would have to have as minimum 3 board names deprecated, the `*_nrf5340_cpuapp`, `*_nrf5340_cpuapp_ns` and `*_nrf5340_cpunet`.

So while I agree that an improve solution is desired for a soc rename would be ideal, then I don't see that a solution with similar limitation as currently exists on zephyr/main should remove the support for renaming a board variant.
imo a variant rename is much more likely / more frequently going to happen than a SoC rename.

Therefore I would have preferred a solution where SoC and variant renames are supported (because they are supported in HWMv1), even if such support has the same limitation as in v1.
But can for now I can accept a solution where just the basic support for v1 -> v2 transition is supported.

> If it were consistent with your overall approach, it would have to be like this:

Correct, if what I wanted to show was all the lines.
My intention was just to show how mapping would be done, not how many lines in `deprecated.cmake` it would require.

In fact I really don't see the number of lines in `deprecated.cmake` as a blocker, cause it's boards (lines) which are anyway going be removed. We could easily populate the file through a script if the gain justifies the work.

Where I think it's important to support matching is the `alias.cmake`, cause that is something a user would want to populate, and there it would be annoying to write 10 lines if a single line can do.